### ISSUE TITLE
Extension methods from ASP.NET Core-related packages do not trigger the C# suggested usings feature in VS

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/roslyn#64712]
Much of the ASP.NET Core API model is based upon the use of extension methods to add and configure features to an application. For those features not in the **Microsoft.AspNetCore.App** shared framework, these extension methods and they're supporting types are in NuGet packages, e.g. [**Microsoft.AspNetCore.Authentication.JwtBearer**](https://nuget.org/packages/Microsoft.AspNetCore.Authentication.JwtBearer):

```csharp
var builder = WebApplication.CreateBuilder();

builder.Services.AddAuthentication()
    // The following extension method is in the Microsoft.AspNetCore.Authentication.JwtBearer NuGet package
    .AddJwtBearer();

var app = builder.Build();

app.MapGet("/", () => "Hello World")
   .RequireAuthorization();

app.Run();
```

The VS feature to suggest usings and add NuGet package references for unresolved symbols provides a convenient way for users to discover and add the required packages to their project when attempting to use an unresolved type. It would be impactful to enable this same functionality for extension methods so that extension methods that add services to the DI container, middleware to the HTTP pipeline, or metadata to an endpoint, can be resolved in a "consume first" fashion.

@davidfowl

**Comment by genlu** [comment]
@CyrusNajmabadi the data that supports this feature is provided by NuGet team, right?

**Comment by CyrusNajmabadi** [comment]
yes.  but we'd have to eevn figure out how to do extension methods.  it's entirely unlikely how normal type lookup works.

**Comment by DamianEdwards** [comment]
Is there any update on this item? Getting support for this would be a great improvement to our developer experience.

**Comment by DamianEdwards** [comment]
Could a more constrained approach make this simpler? e.g. a custom statement completion provider that only targets specific types considered to be "high value", e.g. `IServiceCollection`, etc.

**Comment by CyrusNajmabadi** [comment]
Yes.  That could likely be added to roslyn.  We could give you IVT to create the nuget code-actions in that case.  You'd just need to know the following stuff:

```c#
        [DataMember(Order = 10)]
        public readonly string PackageSource = packageSource;

        [DataMember(Order = 11)]
        public readonly string PackageName = packageName;

        [DataMember(Order = 12)]
        public readonly string PackageVersionOpt = packageVersionOpt;
```

And i think we could take the rest.  We would check if hte package was already installed (and do nothing in that case).

If not installed, we will both recommend isntalling hte same version of that package you might have in other projects, and/or hte recommended package version if you include it, and would finally give the option of using the package manager to install this.  

That seems *Very* reasonable to me.

**Comment by CyrusNajmabadi** [comment]
Looking into a prototype of this.

**Comment by genlu** [comment]
> e.g. a custom statement completion provider that only targets specific types considered to be "high value",

As a complement the codefix, you could implement a completion provider (included in your main framework package) to show those extension methods in completion list when triggered for proper types (if the project doesn't have reference to extension assembly): I think this would make it easier for users to discover those methods (then trigger the codefix) in the first place. Note that there's no way to add reference as part of the completion, which means the completed symbol won't resolve after commit. We currently don't have a precedence of behavior like this, but I think it should be fine as long as there's a codefix to fix it.

**Comment by DamianEdwards** [comment]
@genlu that's an interesting suggestion. Can a custom completion provider specify a custom icon for the suggestion too? That way we could visually differentiate the suggested items that won't actually compile if used unless the code fix action is performed.

**Comment by genlu** [comment]
Yes, that's supported. For example, this code [adds a warning icon](https://github.com/dotnet/roslyn/blob/fb9d3496b655c1aabe1bd587d5ea349f40351f04/src/Features/Core/Portable/Completion/CommonCompletionItem.cs#L41) to the item (although I suspect this is hardcoded to platform-support, might need some fix for a more general usage.) Alternatively, I'd suggest you to simply display a description as text. You could give a warning on the result. This is what we use for displaying extension methods from unimported namespace (the text to the left in the list is the namespace that would be imported)

<img width="405" alt="image" src="https://github.com/dotnet/roslyn/assets/788783/f6eae1c7-fd93-43d0-a76f-c28bec09c767">

**Comment by genlu** [comment]
@DamianEdwards Let me know if you'd like to explore adding a completion provider. We don't have a lot of usage of Roslyn completion provider APIs outside of our repo, despite it being public API. I'd love to work with you to fix any issues you might encounter for implementing one.

**Comment by DamianEdwards** [comment]
@genlu thanks! @captainsafia is going to look at implementing our side of this so we'll reach out to you if we decide to experiment with that approach.
